MoreLinq/Flatten.cs

```diff
@@ -0,0 +1,110 @@
+        /// and all nested sequences (except strings).
+        /// </returns>
+        /// <exception cref="ArgumentNullException"><paramref name="source"/> is null.</exception>
+
+        public static IEnumerable<object> Flatten(this IEnumerable source) =>
+            Flatten(source, obj => !(obj is string));
+
+        /// <summary>
+        /// Flattens a sequence containing arbitrarily-nested sequences. An
+        /// additional parameter specifies a predicate function used to
+        /// determine whether a nested <see cref="IEnumerable"/> should be
+        /// flattened or not.
+        /// </summary>
+        /// <param name="source">The sequence that will be flattened.</param>
+        /// <param name="predicate">
+        /// A function that receives each element that implements
+        /// <see cref="IEnumerable"/> and indicates if its elements should be
+        /// recursively flattened into the resulting sequence.
+        /// </param>
+        /// <returns>
+        /// A sequence that contains the elements of <paramref name="source"/>
+        /// and all nested sequences for which the predicate function
+        /// returned <c>true</c>.
+        /// </returns>
+        /// <exception cref="ArgumentNullException"><paramref name="source"/> is null.</exception>
+        /// <exception cref="ArgumentNullException"><paramref name="predicate"/> is null.</exception>
+
+        public static IEnumerable<object> Flatten(this IEnumerable source, Func<IEnumerable, bool> predicate)
+        {
+            if (source == null) throw new ArgumentNullException(nameof(source));
+            if (predicate == null) throw new ArgumentNullException(nameof(predicate));
+
+            return _(); IEnumerable<object> _()
+            {
+                var e = source.GetEnumerator();
+                var stack = new Stack<IEnumerator>();
+
+                stack.Push(e);
+
+                try
+                {
+                    while (stack.Any())
+                    {
+                        e = stack.Pop();
+
+                        reloop:
+
+                        bool next;
+                        while (next = e.MoveNext())
+                        {
+                            if (e.Current is IEnumerable inner && predicate(inner))
+                            {
+                                stack.Push(e);
+                                e = inner.GetEnumerator();
+                                goto reloop;
+                            }
+                            else
+                            {
+                                yield return e.Current;
+                            }
+                        }
+
+                        if (!next)
+                        {
+                            (e as IDisposable)?.Dispose();
+                        }
+                    }
+                }
+                finally
+                {
+                    stack.Prepend(e)
```

Comment:
The `finally` block should be simple rather than use any fancy LINQ to avoid causing allocations under low memory conditions (the reason for which the `finally` code could be running in the first place). You can re-write this in the same lines of code:

```c#
(e as IDisposable)?.Dispose();
foreach (var se in stack)
    (se as IDisposable)?.Dispose();
```

Note that since [`Stack<>`'s enumerator](https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.stack-1.enumerator?view=netcore-2.0) is a `struct`, there's no cost to iterating over it using `foreach`.